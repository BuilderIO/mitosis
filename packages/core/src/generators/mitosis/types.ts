import { BaseTranspilerOptions } from '@/types/transpiler';

export interface ToMitosisOptions extends BaseTranspilerOptions {
  format: 'react' | 'legacy';
}

export type MitosisMetadata = {};
