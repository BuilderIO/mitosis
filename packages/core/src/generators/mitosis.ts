import dedent from 'dedent';
import json5 from 'json5';
import { format } from 'prettier/standalone';
import { Transpiler } from '..';
import { fastClone } from '../helpers/fast-clone';
import { getComponents } from '../helpers/get-components';
import { getRefs } from '../helpers/get-refs';
import { getStateObjectStringFromComponent } from '../helpers/get-state-object-string';
import { mapRefs } from '../helpers/map-refs';
import { renderPreComponent } from '../helpers/render-imports';
import { METADATA_HOOK_NAME, selfClosingTags } from '../parsers/jsx';
import { MitosisComponent } from '../types/mitosis-component';
import { MitosisNode } from '../types/mitosis-node';
import { blockToReact, componentToReact } from './react';

export interface ToMitosisOptions {
  prettier?: boolean;
  format: 'react' | 'legacy';
}

export const DEFAULT_FORMAT: ToMitosisOptions['format'] = 'legacy';

// Special isValidAttributeName for Mitosis so we can allow for $ in names
const isValidAttributeName = (str: string) => {
  return Boolean(str && /^[$a-z0-9\-_:]+$/i.test(str));
};

export const blockToMitosis = (
  json: MitosisNode,
  toMitosisOptions: Partial<ToMitosisOptions> = {},
): string => {
  const options: ToMitosisOptions = {
    format: DEFAULT_FORMAT,
    ...toMitosisOptions,
  };
  if (options.format === 'react') {
    return blockToReact(json, {
      format: 'lite',
      stateType: 'useState',
      stylesType: 'emotion',
      prettier: options.prettier,
    });
  }

  if (json.name === 'For') {
    const needsWrapper = json.children.length !== 1;
    return `<For each={${json.bindings.each}}>
    {(${json.properties._forName}, index) =>
      ${needsWrapper ? '<>' : ''}
        ${json.children.map((child) => blockToMitosis(child, options))}}
      ${needsWrapper ? '</>' : ''}
    </For>`;
  }

  if (json.properties._text) {
    return json.properties._text as string;
  }

  if (json.bindings._text) {
    return `{${json.bindings._text}}`;
  }

  let str = '';

  str += `<${json.name} `;

  if (json.bindings._spread) {
    str += ` {...(${json.bindings._spread})} `;
  }

  for (const key in json.properties) {
    const value = (json.properties[key] || '')
      .replace(/"/g, '&quot;')
      .replace(/\n/g, '\\n');

    if (!isValidAttributeName(key)) {
      console.warn('Skipping invalid attribute name:', key);
    } else {
      str += ` ${key}="${value}" `;
    }
  }
  for (const key in json.bindings) {
    const value = json.bindings[key] as string;
    if (key === '_spread') {
      continue;
    }

    if (key.startsWith('on')) {
      str += ` ${key}={event => ${value.replace(/\s*;$/, '')}} `;
    } else {
      if (!isValidAttributeName(key)) {
        console.warn('Skipping invalid attribute name:', key);
      } else {
        str += ` ${key}={${value}} `;
      }
    }
  }
  if (selfClosingTags.has(json.name)) {
    return str + ' />';
  }

  // Self close by default if no children
  if (!json.children.length) {
    str += ' />';
    return str;
  }
  str += '>';
  if (json.children) {
    str += json.children
      .map((item) => blockToMitosis(item, options))
      .join('\n');
  }

  str += `</${json.name}>`;

  return str;
};

const getRefsString = (json: MitosisComponent, refs = getRefs(json)) => {
  let str = '';

  for (const ref of Array.from(refs)) {
    str += `\nconst ${ref} = useRef();`;
  }

  return str;
};

const mitosisCoreComponents = ['Show', 'For'];

export const componentToMitosis = (
  toMitosisOptions: Partial<ToMitosisOptions> = {},
): Transpiler => ({ component }) => {
  const options: ToMitosisOptions = {
    format: DEFAULT_FORMAT,
    ...toMitosisOptions,
  };

  if (options.format === 'react') {
    return componentToReact({
      format: 'lite',
      stateType: 'useState',
      stylesType: 'emotion',
      prettier: options.prettier,
    })({ component });
  }

  const json = fastClone(component);

  const refs = getRefs(json);

  mapRefs(json, (refName) => `${refName}.current`);

  const addWrapper = json.children.length !== 1;

  const components = Array.from(getComponents(json));

  const mitosisComponents = components.filter((item) =>
    mitosisCoreComponents.includes(item),
  );
  const otherComponents = components.filter(
    (item) => !mitosisCoreComponents.includes(item),
  );

  const hasState = Boolean(Object.keys(component.state).length);

  const needsMitosisCoreImport = Boolean(
    hasState || refs.size || mitosisComponents.length,
  );

  // TODO: smart only pull in imports as needed
  let str = dedent`
    ${
      !needsMitosisCoreImport
        ? ''
        : `import { ${!hasState ? '' : 'useState, '} ${
            !refs.size ? '' : 'useRef, '
          } ${mitosisComponents.join(', ')} } from '@builder.io/mitosis';`
    }
    ${
      !otherComponents.length
        ? ''
        : `import { ${otherComponents.join(',')} } from '@components';`
    }
    ${renderPreComponent(json)}

    ${
      !component.meta.metadataHook
        ? ''
        : `${METADATA_HOOK_NAME}(${json5.stringify(
            component.meta.metadataHook,
          )})`
    }

    export default function ${component.name}(props) {
      ${
        !hasState
          ? ''
          : `const state = useState(${getStateObjectStringFromComponent(
              json,
            )});`
      }
      ${getRefsString(json)}

      ${!json.hooks.onMount ? '' : `onMount(() => { ${json.hooks.onMount} })`}

      ${
        !json.hooks.onUnMount
          ? ''
          : `onUnMount(() => { ${json.hooks.onUnMount} })`
      }

      return (${addWrapper ? '<>' : ''}
        ${json.children.map((item) => blockToMitosis(item, options)).join('\n')}
        ${addWrapper ? '</>' : ''})
    }

  `;

  if (options.prettier !== false) {
    try {
      str = format(str, {
        parser: 'typescript',
        plugins: [
          require('prettier/parser-typescript'), // To support running in browsers
        ],
      });
    } catch (err) {
      console.error(
        'Format error for file:',
        str,
        JSON.stringify(json, null, 2),
      );
      throw err;
    }
  }
  return str;
};
