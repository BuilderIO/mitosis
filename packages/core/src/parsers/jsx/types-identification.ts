import { types } from '@babel/core';
import { pipe } from 'fp-ts/lib/function';
import { Node, Project, SourceFile, Symbol, SyntaxKind } from 'ts-morph';
import { babelTransformExpression } from '../../helpers/babel-transform';
import { Target } from '../../types/config';
import { mapImportDeclarationToMitosisImport } from './imports';

const getSignalMappingForTarget = (target: Target) => {
  switch (target) {
    case 'svelte':
      const importDeclaration = types.importDeclaration(
        [types.importSpecifier(types.identifier('Writable'), types.identifier('Writable'))],
        types.stringLiteral('svelte/store'),
      );
      importDeclaration.importKind = 'type';

      return {
        getTypeReference: (generics: types.TSType[] = []) =>
          types.tsTypeReference(
            types.identifier('Writable'),
            types.tsTypeParameterInstantiation(generics),
          ),
        importDeclaration,
      };
    default:
      return undefined;
  }
};

export const getSignalMitosisImportForTarget = (target: Target) => {
  const signalType = getSignalMappingForTarget(target);
  if (!signalType) {
    return undefined;
  }
  return mapImportDeclarationToMitosisImport(signalType.importDeclaration);
};

const getSignalSymbol = (project: Project) => {
  const symbolExport = project.createSourceFile(
    'homepage3.lite.tsx',
    `import { Signal } from '@builder.io/mitosis';`,
  );

  // Find the original Signal symbol
  let signalSymbol: Symbol | undefined = undefined;
  symbolExport.forEachDescendant((node) => {
    if (Node.isImportSpecifier(node)) {
      signalSymbol = node.getSymbol()?.getAliasedSymbol();
    }
  });

  if (signalSymbol === undefined) {
    throw new Error('Could not find Signal symbol');
  }
  return signalSymbol as Symbol;
};

export const getSignalImportName = (code: string): string | undefined => {
  let foundSignalUsage = false;
  let signalImportName: string | undefined = undefined;

  babelTransformExpression(code, {
    ImportSpecifier(path) {
      if (types.isIdentifier(path.node.imported) && path.node.imported.name === 'Signal') {
        if (
          path.parentPath.isImportDeclaration() &&
          path.parentPath.node.source.value === '@builder.io/mitosis'
        ) {
          /**
           * in case the import is aliased, we need to use the local name,
           * e.g. `import { Signal as MySignal } from '@builder.io/mitosis'`
           */
          signalImportName = path.node.local.name;
          path.stop();
        }
      }
    },
  });

  if (!signalImportName) {
    return undefined;
  }

  babelTransformExpression(code, {
    TSTypeReference(path) {
      if (types.isIdentifier(path.node.typeName) && path.node.typeName.name === signalImportName) {
        foundSignalUsage = true;
        path.stop();
      }
    },
  });

  return foundSignalUsage ? signalImportName : undefined;
};

const addSignalImport = ({ code, target }: { code: string; target: Target }) => {
  const signalType = getSignalMappingForTarget(target);

  if (!signalType) {
    return code;
  }

  return babelTransformExpression(code, {
    Program(path) {
      path.node.body.unshift(signalType.importDeclaration);
    },
  });
};
/**
 * Finds all `Signal` types and replaces them with the correct type for the given target.
 * e.g. `Signal<string>` becomes `Writable<string>` for Svelte.
 */
export const mapSignalType = ({
  code,
  target,
  signalImportName = getSignalImportName(code),
}: {
  code: string;
  target: Target;
  signalImportName?: string;
}) => {
  const signalType = getSignalMappingForTarget(target);

  return babelTransformExpression(code, {
    TSTypeReference(path) {
      if (types.isIdentifier(path.node.typeName) && path.node.typeName.name === signalImportName) {
        const params = path.node.typeParameters?.params || [];

        const newType = signalType?.getTypeReference
          ? signalType.getTypeReference(params)
          : // if no mapping exists, drop `Signal` and just use the generic type passed to `Signal` as-is.
            params[0];

        path.replaceWith(newType);
      }
    },
  });
};

/**
 * Processes the `Signal` type usage in a plain TS file:
 * - Finds the Signal import name
 * - Maps the Signal type to the target's equivalent
 * - Adds the equivalent of the Signal import to the file
 */
export const mapSignalTypeInTSFile = ({ code, target }: { code: string; target: Target }) => {
  const signalImportName = getSignalImportName(code);

  if (!signalImportName) {
    return code;
  }

  return pipe(mapSignalType({ target, code, signalImportName }), (code) =>
    addSignalImport({ code, target }),
  );
};

const getProject = (tsConfigFilePath: string) => {
  try {
    return new Project({ tsConfigFilePath });
  } catch (err) {
    throw new Error(
      'Error creating Typescript Project. Make sure `tsConfigFilePath` points to a valid tsconfig.json file',
    );
  }
};

export const createTypescriptProject = (tsConfigFilePath: string) => {
  const project = getProject(tsConfigFilePath);
  const signalSymbol = getSignalSymbol(project);
  return { project, signalSymbol };
};

const getPropsSymbol = (ast: SourceFile) => {
  let propsSymbol: Symbol | undefined = undefined;
  ast.forEachChild((node) => {
    if (propsSymbol !== undefined) return;

    if (Node.isArrowFunction(node) || Node.isFunctionDeclaration(node)) {
      if (
        node.hasModifier(SyntaxKind.ExportKeyword) &&
        node.hasModifier(SyntaxKind.DefaultKeyword)
      ) {
        propsSymbol = node.getParameters()[0]?.getSymbol();
      }
    }
  });

  return propsSymbol as Symbol | undefined;
};

const getContextSymbols = (ast: SourceFile) => {
  const contextSymbols = new Set<Symbol>();

  ast.forEachDescendant((node) => {
    if (!Node.isVariableDeclaration(node)) return;

    const initializer = node.getInitializer();

    if (!Node.isCallExpression(initializer)) return;
    if (initializer.getExpression().getText() !== 'useContext') return;

    const contextSymbol = node.getNameNode().getSymbol();

    if (contextSymbol === undefined) return;

    contextSymbols.add(contextSymbol);
  });

  return contextSymbols;
};

export const findSignals = (args: {
  project: Project;
  signalSymbol: Symbol;
  code?: string;
  filePath?: string;
}) => {
  const { project, signalSymbol } = args;

  const ast = args.code
    ? args.project.createSourceFile('homepage2.lite.tsx', args.code)
    : args.filePath
    ? args.project.getSourceFileOrThrow(args.filePath)
    : undefined;

  if (ast === undefined) {
    throw new Error('Could not find AST. Please provide either `code` or `filePath` configs.');
  }

  const reactiveValues = {
    props: new Set<string>(),
    state: new Set<string>(),
    context: new Set<string>(),
  };

  const propsSymbol = getPropsSymbol(ast);

  const contextSymbols = getContextSymbols(ast);

  ast.forEachDescendant((parentNode) => {
    if (Node.isPropertyAccessExpression(parentNode)) {
      const node = parentNode.getExpression();
      const aliasSymbol = node.getType().getTargetType()?.getAliasSymbol();
      const isSignal = aliasSymbol === signalSymbol;

      if (!isSignal) return;

      let isInsideType = false;
      let isInsideDeclaration = false;
      node.getParentWhile((parent, child) => {
        // stop once we hit the function block
        if (Node.isBlock(child) || Node.isBlock(parent)) {
          return false;
        }

        // crawl up parents to make sure we're not inside a type
        if (Node.isTypeNode(parent) || Node.isTypeAliasDeclaration(parent)) {
          isInsideType = true;
          return false;
        }

        return true;
      });

      if (isInsideType) return;
      if (isInsideDeclaration) return;

      const nodeSymbol = node.getSymbol();

      if (
        Node.isPropertyAccessExpression(node) &&
        node.getExpression().getSymbol() === propsSymbol
      ) {
        reactiveValues.props.add(node.getNameNode().getText());
      } else if (nodeSymbol && contextSymbols.has(nodeSymbol)) {
        reactiveValues.context.add(node.getText());
      } else {
        reactiveValues.state.add(node.getText());
      }
    }
  });

  return reactiveValues;
};
