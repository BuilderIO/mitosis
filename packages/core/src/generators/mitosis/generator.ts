import { HOOKS } from '@/constants/hooks';
import { SELF_CLOSING_HTML_TAGS } from '@/constants/html_tags';
import { ToMitosisOptions } from '@/generators/mitosis/types';
import { dedent } from '@/helpers/dedent';
import { checkIsEvent } from '@/helpers/event-handlers';
import { fastClone } from '@/helpers/fast-clone';
import { getComponents } from '@/helpers/get-components';
import { getRefs } from '@/helpers/get-refs';
import { getStateObjectStringFromComponent } from '@/helpers/get-state-object-string';
import { isMitosisNode } from '@/helpers/is-mitosis-node';
import { isRootTextNode } from '@/helpers/is-root-text-node';
import { mapRefs } from '@/helpers/map-refs';
import { renderPreComponent } from '@/helpers/render-imports';
import { checkHasState } from '@/helpers/state';
import {
  runPostCodePlugins,
  runPostJsonPlugins,
  runPreCodePlugins,
  runPreJsonPlugins,
} from '@/modules/plugins';
import { MitosisComponent } from '@/types/mitosis-component';
import { MitosisNode, checkIsForNode, checkIsShowNode } from '@/types/mitosis-node';
import { TranspilerGenerator } from '@/types/transpiler';
import json5 from 'json5';
import { check, format } from 'prettier/standalone';

import { blockToReact, componentToReact } from '../react';

export const DEFAULT_FORMAT: ToMitosisOptions['format'] = 'legacy';

// Special isValidAttributeName for Mitosis so we can allow for $ in names
const isValidAttributeName = (str: string) => {
  return Boolean(str && /^[$a-z0-9\-_:]+$/i.test(str));
};

const isInvalidJsxAttributeName = (str: string) => {
  let attr = str.trim();
  if (attr.startsWith(':') || str.startsWith('@')) {
    return true;
  }
  return false;
};

export const blockToMitosis = (
  json: MitosisNode,
  toMitosisOptions: Partial<ToMitosisOptions> = {},
  component: MitosisComponent,
  insideJsx: boolean,
): string => {
  const options: ToMitosisOptions = {
    format: DEFAULT_FORMAT,
    ...toMitosisOptions,
  };
  if (options.format === 'react') {
    return blockToReact(
      json,
      {
        format: 'lite',
        stateType: 'useState',
        stylesType: 'emotion',
        type: 'dom',
        prettier: options.prettier,
      },
      component,
      insideJsx,
    );
  }

  if (checkIsShowNode(json)) {
    const when = json.bindings.when?.code;
    const elseCase = json.meta.else as MitosisNode;
    if (options.nativeConditionals) {
      const needsWrapper = json.children.length !== 1;

      const renderChildren = `${needsWrapper ? '<>' : ''}
        ${json.children
          .map((child) => blockToMitosis(child, options, component, needsWrapper))
          .join('\n')}
    ${needsWrapper ? '</>' : ''}`;

      const renderElse =
        elseCase && isMitosisNode(elseCase)
          ? blockToMitosis(elseCase, options, component, false)
          : 'null';
      return `${insideJsx ? '{' : ''}(${when}) ? ${renderChildren} : ${renderElse}${
        insideJsx ? '}' : ''
      }`;
    } else {
      const elseHandler = elseCase
        ? ` else={${blockToMitosis(elseCase, options, component, false)}}`
        : '';
      return `<Show when={${when}}${elseHandler}>
  ${json.children.map((child) => blockToMitosis(child, options, component, true)).join('\n')}
</Show>`;
    }
  }

  if (checkIsForNode(json)) {
    const needsWrapper = json.children.length !== 1;
    if (options.nativeLoops) {
      const a = `${insideJsx ? '{' : ''}(${json.bindings.each?.code}).map(
      (${json.scope.forName || '_'}, ${json.scope.indexName || 'index'}) => (
      ${needsWrapper ? '<>' : ''}
        ${json.children
          .map((child) => blockToMitosis(child, options, component, needsWrapper))
          .join('\n')}
      ${needsWrapper ? '</>' : ''}
      ))${insideJsx ? '}' : ''}`;
      return a;
    }
    return `<For each={${json.bindings.each?.code}}>
    {(${json.scope.forName || '_'}, ${json.scope.indexName || 'index'}) =>
      ${needsWrapper ? '<>' : ''}
        ${json.children.map((child) => blockToMitosis(child, options, component, needsWrapper))}}
      ${needsWrapper ? '</>' : ''}
    </For>`;
  }

  if (json.properties._text) {
    let text = json.properties._text
      // Convert HTML <br> to JSX-valid <br />
      .replace(/<br\s*>/g, '<br />');

    let isInvalidJsx = text.includes('{') || text.includes('}');

    if (text.includes('<') || text.includes('>')) {
      // test if this can parse as jsx
      try {
        /**
         * We intentionally use the typescript parser here because texts like ">" will crash
         * in the typescript parser but will not crash in the babel parser. The Prettier
         * formatting that is run after JSX is generated also uses the typescript parser,
         * so we want to make sure that doesn't crash.
         */
        check(`let _ = <>${text}</>;`, {
          parser: 'typescript',
          plugins: [
            require('prettier/parser-typescript'), // To support running in browsers
          ],
        });
        isInvalidJsx = false;
      } catch (e) {
        isInvalidJsx = true;
      }
    }

    if (isInvalidJsx) {
      text = text
        .replace(/>/g, '&gt;')
        .replace(/</g, '&lt;')
        .replace(/{/g, '&#123;')
        .replace(/}/g, '&#125;')
        .replace(/&/g, '&amp;');
    }

    if (insideJsx) {
      return `${text}`;
    } else {
      return `<>${text}</>`;
    }
  }

  if (json.bindings._text?.code) {
    if (insideJsx) {
      return `{${json.bindings._text.code}}`;
    } else {
      return `${json.bindings._text.code}`;
    }
  }

  let str = '';

  str += `<${json.name} `;

  for (const key in json.properties) {
    if (isInvalidJsxAttributeName(key)) {
      console.warn('Skipping invalid attribute name:', key);
      continue;
    }
    const value = (json.properties[key] || '').replace(/"/g, '&quot;').replace(/\n/g, '\\n');

    if (!isValidAttributeName(key)) {
      console.warn('Skipping invalid attribute name:', key);
    } else {
      str += ` ${key}="${value}" `;
    }
  }
  for (const key in json.bindings) {
    if (isInvalidJsxAttributeName(key)) {
      console.warn('Skipping invalid attribute name:', key);
      continue;
    }
    const value = json.bindings[key]?.code as string;

    if (!value || json.slots?.[key]) {
      continue;
    }

    if (json.bindings[key]?.type === 'spread') {
      str += ` {...(${json.bindings[key]?.code})} `;
    } else if (checkIsEvent(key)) {
      const { arguments: cusArgs = ['event'], async } = json.bindings[key]!;
      const asyncKeyword = async ? 'async ' : '';
      str += ` ${key}={${asyncKeyword}(${cusArgs.join(',')}) => ${value.replace(/\s*;$/, '')}} `;
    } else {
      if (!isValidAttributeName(key)) {
        console.warn('Skipping invalid attribute name:', key);
      } else {
        str += ` ${key}={${value}} `;
      }
    }
  }

  for (const key in json.slots) {
    const value = json.slots[key];
    str += ` ${key}={`;
    if (value.length > 1) {
      str += '<>';
    }
    str += json.slots[key]
      .map((item) => blockToMitosis(item, options, component, insideJsx))
      .join('\n');
    if (value.length > 1) {
      str += '</>';
    }
    str += `}`;
  }

  if (SELF_CLOSING_HTML_TAGS.has(json.name)) {
    return str + ' />';
  }

  // Self close by default if no children
  if (!json.children.length) {
    str += ' />';
    return str;
  }
  str += '>';

  if (json.children) {
    str += json.children.map((item) => blockToMitosis(item, options, component, true)).join('\n');
  }

  str += `</${json.name}>`;

  return str;
};

const getRefsString = (json: MitosisComponent, refs = Array.from(getRefs(json))) => {
  let str = '';

  for (const ref of refs) {
    const typeParameter = json['refs'][ref]?.typeParameter || '';
    const argument = json['refs'][ref]?.argument || '';
    str += `\nconst ${ref} = useRef${typeParameter ? `<${typeParameter}>` : ''}(${argument});`;
  }

  return str;
};

const mitosisCoreComponents = ['Show', 'For'];

export const componentToMitosis: TranspilerGenerator<Partial<ToMitosisOptions>> =
  (toMitosisOptions = {}) =>
  ({ component }) => {
    const options: ToMitosisOptions = {
      format: DEFAULT_FORMAT,
      ...toMitosisOptions,
    };

    if (options.format === 'react') {
      return componentToReact({
        format: 'lite',
        stateType: 'useState',
        stylesType: 'emotion',
        prettier: options.prettier,
      })({ component });
    }

    let json = fastClone(component);

    if (options.plugins) {
      json = runPreJsonPlugins({ json, plugins: options.plugins });
    }

    const domRefs = getRefs(component);
    // grab refs not used for bindings
    const jsRefs = Object.keys(component.refs).filter((ref) => domRefs.has(ref));

    const refs = [...jsRefs, ...Array.from(domRefs)];

    mapRefs(json, (refName) => {
      return `${refName}${domRefs.has(refName) ? `.current` : ''}`;
    });

    const addWrapper = json.children.length !== 1 || isRootTextNode(json);

    const components = Array.from(getComponents(json));
    const mitosisCoreComponents: string[] = [];
    if (!options.nativeConditionals) {
      mitosisCoreComponents.push('Show');
    }
    if (!options.nativeLoops) {
      mitosisCoreComponents.push('For');
    }

    const mitosisComponents = components.filter((item) => mitosisCoreComponents.includes(item));
    const otherComponents = components.filter((item) => !mitosisCoreComponents.includes(item));

    if (options.plugins) {
      json = runPostJsonPlugins({ json, plugins: options.plugins });
    }

    const hasState = checkHasState(component);

    const needsMitosisCoreImport = Boolean(hasState || refs.length || mitosisComponents.length);

    const stringifiedUseMetadata = json5.stringify(component.meta.useMetadata);

    // TODO: smart only pull in imports as needed
    let str = dedent`
    ${
      !needsMitosisCoreImport
        ? ''
        : `import { ${!hasState ? '' : 'useStore, '} ${
            !refs.length ? '' : 'useRef, '
          } ${mitosisComponents.join(', ')} } from '@builder.io/mitosis';`
    }
    ${!otherComponents.length ? '' : `import { ${otherComponents.join(',')} } from '@components';`}
    ${json.types ? json.types.join('\n') : ''}

    ${renderPreComponent({
      explicitImportFileExtension: options.explicitImportFileExtension,
      component: json,
      target: 'mitosis',
    })}

    ${
      stringifiedUseMetadata && stringifiedUseMetadata !== '{}'
        ? `${HOOKS.METADATA}(${stringifiedUseMetadata})`
        : ''
    }

    export default function ${component.name}(props) {
      ${!hasState ? '' : `const state = useStore(${getStateObjectStringFromComponent(json)});`}
      ${getRefsString(json, refs)}

      ${json.hooks.onMount.map((hook) => `onMount(() => { ${hook.code} })`)}

      ${!json.hooks.onUnMount?.code ? '' : `onUnMount(() => { ${json.hooks.onUnMount.code} })`}

      ${json.style ? `useStyle(\`${json.style}\`)` : ''}

      return ${options.returnArray ? '[' : '('}${addWrapper ? '<>' : ''}
        ${json.children
          .map((item) => blockToMitosis(item, options, component, addWrapper))
          .join('\n')}
        ${addWrapper ? '</>' : ''}${options.returnArray ? ']' : ')'}
    }

  `;

    if (options.plugins) {
      str = runPreCodePlugins({ json, code: str, plugins: options.plugins });
    }
    if (options.prettier !== false) {
      try {
        str = format(str, {
          parser: 'typescript',
          plugins: [
            require('prettier/parser-typescript'), // To support running in browsers
          ],
        });
      } catch (err) {
        if (process.env.NODE_ENV !== 'test') {
          console.error('Format error for file:', str, JSON.stringify(json, null, 2));
        }
        throw err;
      }
    }
    if (options.plugins) {
      str = runPostCodePlugins({ json, code: str, plugins: options.plugins });
    }
    return str;
  };
