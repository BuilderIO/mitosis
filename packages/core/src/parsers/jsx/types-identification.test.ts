import { createTypescriptProject, findSignals, mapSignalType } from './types-identification';

const code = `
import { Signal, useState, useContext, createContext } from '@builder.io/mitosis';

const FooContext = createContext({ foo: 'bar' }, { reactive: true });
const NormalContext = createContext({ foo: 'bar' });

type K = Signal<string>;

type Props = {
  k: K;
  another: Signal<number>
};
    
export default function MyComponent(props: Props) {
  const [n] = useState(123, { reactive: true });

  const context = useContext(FooContext)
  const normalContext = useContext(NormalContext)

  console.log(
    n, 
    n.value, 
    props.k, 
    props.k.value,
    context,
    context.value.foo,
    normalContext,
    normalContext.value.foo,
    );

  return <RenderBlock 
    a={props.k} 
    b={props.k.value} 
    c={n} 
    d={n.value} 
    e={context}
    f={context.value.foo}
    g={normalContext}
    h={normalContext.value.foo}
    />;
};
`;
describe('Signals type parsing', () => {
  test(findSignals.name, () => {
    const result = findSignals({
      code,
      // we piggyback on the e2e-app TS project to avoid having to setup one for this test.
      ...createTypescriptProject(__dirname + '/../../../../../e2e/e2e-app/tsconfig.json'),
    });
    expect(result).toMatchSnapshot();
  });

  describe(mapSignalType.name, () => {
    test('svelte', () => {
      const result = mapSignalType({
        target: 'svelte',
        code,
        // we piggyback on the e2e-app TS project to avoid having to setup one for this test.
        ...createTypescriptProject(__dirname + '/../../../../../e2e/e2e-app/tsconfig.json'),
      });
      expect(result).toMatchSnapshot();
    });

    test('react', () => {
      const result = mapSignalType({
        target: 'react',
        code,
        // we piggyback on the e2e-app TS project to avoid having to setup one for this test.
        ...createTypescriptProject(__dirname + '/../../../../../e2e/e2e-app/tsconfig.json'),
      });
      expect(result).toMatchSnapshot();
    });
  });
});
