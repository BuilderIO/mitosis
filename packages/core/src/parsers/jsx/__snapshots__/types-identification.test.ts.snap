// Vitest Snapshot v1

exports[`findSignals > x 1`] = `
{
  "context": Set {
    "context",
  },
  "props": Set {
    "k",
  },
  "state": Set {
    "n",
  },
}
`;
